File: sorting-algorithms/bubble-sort/visual-basic/bubble_sort_version_1.vb
' Raspberry Pi Foundation
' Developed as part of Ada Computer Science
' Usage licensed under CC BY-NC-SA 4.0
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program

Module Program

    ' The Main subroutine is the default entry point for a VB program
    Sub Main()
        Dim testItems As Integer() = {43, 21, 2, 50, 3, 80, 35, 7, 64} 
        'Dim testItems As Integer() = {80, 64, 50, 43, 35, 21, 7, 3, 2}; 'Least sorted
        'Dim testItems As Integer() = {2, 3, 7, 35, 43, 21, 50, 64, 80}; 'Nearly sorted
        'Dim testItems As Integer() = {2, 3, 7, 21, 35, 43, 50, 64, 80}; 'Sorted
        'Dim testItems As Integer() = {43, 21, 2, 50, 3, 80, 35, 7, 64}; 'Random
        Console.WriteLine("### Bubble sort version 1 (for loop) ###")
        Console.WriteLine("[{0}]", String.Join(", ", testItems))  ' Display array of test data
        BubbleSortVersion1(testItems)
    End Sub

    Sub BubbleSortVersion1(ByVal items As Integer())
        Dim numItems As Integer = items.Length

        For passNum As Integer = 1 To numItems - 1

            For index As Integer = 0 To numItems - 2

                If items(index) > items(index + 1) Then
                    Dim temp As Integer = items(index)
                    items(index) = items(index + 1)
                    items(index + 1) = temp
                End If
            Next

            Console.WriteLine("[{0}]", String.Join(", ", items))
        Next
    End Sub

End Module

File: sorting-algorithms/bubble-sort/java/bubble_sort_version_1.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science

Usage licensed under CC BY-NC-SA 4.0
 
Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

import java.util.Arrays;

class SortingAlgorithms {
    // The main method is the entry point for all Java programs
    public static void main(String[] args) {
        // Perform a bubble sort on the test data
        //int[] testItems = new int[] {80, 64, 50, 43, 35, 21, 7, 3, 2}; // Least sorted
        //int[] testItems = new int[] {2, 3, 7, 35, 43, 21, 50, 64, 80}; // Nearly sorted
        //int[] testItems = new int[] {2, 3, 7, 21, 35, 43, 50, 64, 80}; // Sorted
        int[] testItems = new int[] {43, 21, 2, 50, 3, 80, 35, 7, 64}; // Random

        System.out.println("### Bubble sort version 1 (for loops) ###");
        System.out.println(Arrays.toString(testItems));

        bubbleSortVersion1(testItems);
    }


    // An inefficient bubble sort that uses nested for loops
    public static void bubbleSortVersion1(int[] items) {
        // Initialise the variables
        int numItems = items.length;

        // Pass through the array of items n-1 times
        for (int passNum = 1; passNum < numItems; passNum++) {
            // Perform a pass
            for (int index = 0; index < numItems - 1; index++) {
                // Compare items to check if they are out of order
                if (items[index] > items[index + 1]) {
                    // Swap the items
                    int temp = items[index];
                    items[index] = items[index + 1];
                    items[index + 1] = temp;
                }
            }
            System.out.println(Arrays.toString(items)); // Testing
        }
    }


}